src/scheduler/Executor.cpp

```diff
@@ -201,11 +411,58 @@ void Executor::executeTasks(std::vector<int> msgIdxs,
     }
 }
 
+std::shared_ptr<faabric::util::SnapshotData> Executor::getMainThreadSnapshot(
+  faabric::Message& msg,
+  bool createIfNotExists)
+{
+    std::string snapshotKey = faabric::util::getMainThreadSnapshotKey(msg);
+    bool exists = false;
```

Comment:
Again, make `exists` a `std::atomic<bool>` and access it using `exists.load(std::memory_order_acquire)`